### PR TITLE
MGMT-9778: align authorization for events

### PR DESCRIPTION
**Pull request by slaviered**
The event handler implements its access checks within the application layer. These verifications do not conform to the new tenancy policies.

This PR re-implements the access restrictions using OwnedBy API and corresponding SQL queries that conform to the new API.

It also aligns the List behavior of events with the standard requirements for list behavior.

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/events/event.go [hunk 1]
```diff
@@ -176,85 +177,67 @@ func (e *Events) V2AddMetricsEvent(ctx context.Context, clusterID *strfmt.UUID,
...
-					"user %s is not authorized to query events for infra_env %s: infra_env owned by another user.",
-					user, infraEnvID.String())
-				return &gorm.DB{Error: errors.Wrapf(gorm.ErrRecordNotFound, "infra_env %s not found.", infraEnvID.String())}
-			}
-		} else { // A case where there is no user in context, simply filter by resource id
-			whereCondition = append(whereCondition, fmt.Sprintf("infra_env_id = '%s'", infraEnvID.String()))
-		}
+
+	//prepare the common parts of the query
+	db := e.db.Debug().Order("event_time").Where("category IN (?)", selectedCategories)
```
Comment (danielerez) [inline internal/events/event.go:196]: do we need the debug log here?

File: internal/events/event.go [hunk 2]
```diff
@@ -176,85 +177,67 @@ func (e *Events) V2AddMetricsEvent(ctx context.Context, clusterID *strfmt.UUID,
...
-	if len(whereCondition) == 0 {
-		queryErr := errors.Wrap(gorm.ErrInvalidTransaction, "events query with no filters is not allowed.")
-		e.log.Error(queryErr)
-		return &gorm.DB{Error: queryErr}
+
+	var result *gorm.DB
+
+	//retrieveing all events can be done only by admins. This is done to restrict data
+	//intensive queries by common users
+	if clusterID == nil && infraEnvID == nil && hostID == nil && identity.IsAdmin(ctx) {
```
Comment (danielerez) [inline internal/events/event.go:205]: maybe worth having an helper func for 'clusterID == nil && infraEnvID == nil && hostID == nil' - something like isAllEventsRequested.

File: internal/events/event.go (hunk 1 above)
Comment (slaviered) [reply internal/events/event.go:196]: Left by mistake. I will remove it

File: internal/events/event.go (hunk 2 above)
Comment (slaviered) [reply internal/events/event.go:205]: OK. I will add helpers for all the cases

File: internal/events/event_test.go [hunk 3]
```diff
@@ -205,390 +208,132 @@ var _ = Describe("Events library", func() {
...
+				Expect(err).ShouldNot(HaveOccurred())
+				Expect(len(evs)).To(Equal(2))
+			})
+
+			It("gets infra-env's events when specifying infra-env", func() {
+				evs, err := theEvents.V2GetEvents(ctx, nil, nil, &infraEnv1)
+				Expect(err).ShouldNot(HaveOccurred())
+				Expect(len(evs)).To(Equal(3))
+			})
+			It("gets hosts's events when specifying host", func() {
```
Comment (nmagnezi) [inline internal/events/event_test.go:264]: Should we have a test to fetch events by host_id alone?

Comment (nmagnezi) [inline internal/events/event_test.go:264]: Should we have a test to fetch events by host_id alone?

File: internal/events/event_test.go [hunk 4]
```diff
@@ -205,390 +208,132 @@ var _ = Describe("Events library", func() {
...
+				payload.Organization = "org1"
+				ctx = context.WithValue(context.TODO(), restapi.AuthKey, payload)
+			})
+			It("gets events on own clusters", func() {
+				By("strictly own cluster")
+				evs, err := theEvents.V2GetEvents(ctx, &cluster1, nil, nil)
+				Expect(err).ShouldNot(HaveOccurred())
+				Expect(len(evs)).To(Equal(2))
+
+				By("cluster on the same org")
```
Comment (nmagnezi) [inline internal/events/event_test.go:292]: the same org gets 1 event where the actual owner gets 2? which event got filtered out?

Comment (nmagnezi) [inline internal/events/event_test.go:292]: the same org gets 1 event where the actual owner gets 2? which event got filtered out?

**Review by nmagnezi — CHANGES_REQUESTED** [review]
Looks great overall, just a few minor comments.
One thing missing here is coverage with subsystem tests.

File: internal/events/event.go (hunk 2 above)
Comment (nmagnezi) [reply internal/events/event.go:205]: Should we keep a debug log in place for this case?

File: internal/events/event_test.go (hunk 3 above)
Comment (slaviered) [reply internal/events/event_test.go:264]: added

File: internal/events/event_test.go (hunk 4 above)
Comment (slaviered) [reply internal/events/event_test.go:292]: It's not the same cluster. on cluster 1 there are also infra-env events while on cluster 2 there's only 1 event

Comment (slaviered) [reply internal/events/event_test.go:292]: I have added an examination of the returned events so now the tests also document the API behavior

**Pull request merged**